src/devices/Lcm1602a1/HD44780.cs

```diff
@@ -0,0 +1,559 @@
+        /// 5x8 character. The underline cursor just turns on the entire bottom row.
+        /// 
+        /// 5x10 mode is effectively useless as displays aren't available that utilize it. In 5x10
+        /// mode *16* bytes of data are used for each character. That leaves room for only *4*
+        /// custom characters. The first character is addressable from code 0, 1, 8, and 9. The
+        /// second is 2, 3, 10, 11 and so on...
+        /// 
+        /// In this mode *11* bytes of data are actually used for the character data, which
+        /// effectively gives you a 5x11 character, although typically the last line is blank to
+        /// leave room for the underline cursor. Why the modes are referred to as 5x8 and 5x10 as
+        /// opposed to 5x7 and 5x10 or 5x8 and 5x11 is a mystery. In an early pre-release data
+        /// book 5x7 and 5x10 is used (Advance Copy #AP4 from July 1985). Perhaps it was a
+        /// marketing change?
+        /// 
+        /// As only 11 bytes are used in 5x10 mode, but 16 bytes are reserved, the last 5 bytes
+        /// are useless. The datasheet helpfully suggests that you can store your own data there.
+        /// The same would be true for bits 5-7 of lines that matter for both 5x8 and 5x10.
+        /// </remarks>
+        /// <param name="location">Should be between 0 and 7</param>
+        /// <param name="characterMap">Provide an array of 8 bytes containing the pattern</param>
+        public void CreateCustomCharacter(byte location, params byte[] characterMap)
+        {
+            if (location > 7)
+                throw new ArgumentOutOfRangeException(nameof(location));
+
+            if (characterMap.Length != 8)
+                throw new ArgumentException(nameof(characterMap));
+
+            Send((byte)(SetCGRamAddressCommand | (location << 3)));
+
+            for (int i = 0; i < 8; i++)
+            {
+                Send(characterMap[i], data: true);
+            }
+        }
+
+        /// <summary>
+        /// Write text to display.
+        /// </summary>
+        /// <remarks>
+        /// There are only 256 characters available. There are chip variants
+        /// with different character sets. Characters from space ' ' (32) to
+        /// '}' are usually the same with the exception of '\', which is a
+        /// yen symbol on some chips '¥'.
+        /// </remarks>
+        /// <param name="value">Text to be displayed.</param>
+        public void Write(string value)
+        {
+            for (int i = 0; i < value.Length; ++i)
+            {
+                Send((byte)value[i], data: true);
+            }
+        }
+
+        /// <summary>
+        /// Send a data or command byte to the controller.
+        /// </summary>
+        /// <param name="data">True to send data, otherwise sends a command.</param>
+        private void Send(byte value, bool data = false)
+        {
+            _controller.Write(_rsPin, data ? PinValue.High : PinValue.Low);
```

Comment:
@joperezr another instance where bool overload would be useful :smile:

Reply:
Referencing #121

Reply:
Or int for that matter. I wonder if we'd be better off making PinValue a struct. Then we could easily add implicit conversions.

``` C#
    class Program
    {
        static void Main(string[] args)
        {
            PinValue value = true;
            Console.WriteLine($"True is {value}");
            TestApi(0);
            Console.WriteLine($"Equality is {PinValue.High == true}");
        }

        public static void TestApi(PinValue value)
        {
            Console.WriteLine($"PinValue is {value}");
        }
    }

    public readonly struct PinValue : IEquatable<PinValue>
    {
        // int so the struct is blittable (anything other than bool)
        private readonly int _value;

        private PinValue(int value) => _value = value;

        public static PinValue High => new PinValue(1);
        public static PinValue Low => new PinValue(0);

        public bool Equals(PinValue other) => other._value == _value;

        public static implicit operator PinValue(int value) => value == 0 ? Low : High;
        public static implicit operator PinValue(bool value) => value ? High : Low;

        public static bool operator ==(PinValue a, PinValue b) => a.Equals(b);
        public static bool operator !=(PinValue a, PinValue b) => !a.Equals(b);

        public override string ToString() => _value == 0 ? "Low" : "High";
    }
```